Draft for an event with all information after schema change was perfomed

beberlei opened the pull request:
/cc @dbu 

Related:
* https://github.com/symfony/symfony/issues/54039

beberlei committed:
Draft for an event with all information after schema change was performed.

dbu commented on src/Tools/SchemaTool.php:
```diff
@@ -83,6 +86,12 @@ public function createSchema(array $classes): void
+        $eventManager = $this->em->getEventManager();
+        $eventManager->dispatchEvent(
+            ToolEvents::postSchemaChanged,
+            new SchemaChangedEventArgs($this->em, $schema, new Schema(), $createSchemaSql),
```
thank you for looking into this. so the idea would be that e.g. symfony messsenger could listen to this, inspect the changes and fire its own sql statements against the connection if it sees that something was done which it wants to alter?

is this event also useful when doctrine migrations are used rather than database updated on the fly?

with dbal 3, symfony was adding extra sql in https://github.com/symfony/symfony/blob/73f8713e143c64dc5f045dcdee7d5e893cccfcd8/src/Symfony/Bridge/Doctrine/SchemaListener/MessengerTransportDoctrineSchemaListener.php#L76C1-L76C38 (and [that sql](https://github.com/symfony/symfony/blob/73f8713e143c64dc5f045dcdee7d5e893cccfcd8/src/Symfony/Component/Messenger/Bridge/Doctrine/Transport/PostgreSqlConnection.php#L115) afaik can not be represented in the doctrine schema models.
that sql got saved to the migration file, which was quite nice to review it.

beberlei replied:
Yes, something like:

```php
public function onPostSchemaChanged($event)
{
    if (!$this->hasMessengerTable($event)) {
        $event->getEntityManage()->getConnection()->executeStatement($this->getMessengerTableSql());
    }
}
```

dbu replied:
and how can this be integrated with the doctrine migrations? could the listener also add sql into a migration?

greg0ire commented:
Cc @wmouwen

beberlei committed:
Housekeeping: phpcs

beberlei commented:
Just to make it clear, I would greenlight this approach, but I need someone else to take over and finish this:

* getters on the event
* tests
* integrate into dropSchema/getDropSchemaSql

greg0ire committed:
Add getters on the event

greg0ire committed:
Document the new event

greg0ire committed:
Demonstrate that listener is called

For now, that test only works with DBAL 3, but hopefully that's only
because of the test mocking system rather than the code.

Perf commented:
> Just to make it clear, I would greenlight this approach, but I need someone else to take over and finish this:
> 
> * [x]  getters on the event
> * [ ]  tests
> * [ ]  integrate into dropSchema/getDropSchemaSql
> * [x]  documentation
> * [ ]  end to end test: Does this really work when executing multiple times

hey @greg0ire, thank you for reopening this PR!
I would be glad to help, but not sure if I would be helpful, I have basic knowledge about doctrine/orm internals.
If you or someone else will provide me more tech details on what needs to be done, probably, I could contribute.

greg0ire commented:
Closing as per https://github.com/symfony/symfony/issues/54039#issuecomment-2827990060

@Perf thanks for your proposal. A good first step would be to address https://github.com/symfony/symfony/issues/54039#issuecomment-2827991790
